#3: Installed slc dependencies via ansible

tomuben opened the pull request:
fixes #3

tomuben committed:
#3: Installed slc dependencies via ansible

tkilias commented on exasol_script_languages_developer_sandbox/ansible/install_docker.yml:
```diff
@@ -0,0 +1,40 @@
+#- name: Install aptitude
```
Do we need this?

tomuben replied:
Removed

tkilias commented on exasol_script_languages_developer_sandbox/lib/ansible/ansible_repository.py:
```diff
@@ -0,0 +1,39 @@
+        env = jinja2.Environment(loader=loader, autoescape=jinja2.select_autoescape(), keep_trailing_newline=True)
+        for ansible_file in loader.list_templates():
+            t = env.get_template(ansible_file)
+            content = t.render()
```
You use jinka2 to load the file, but call render without variables. What is the intention? And if this includes Ansible files, this might be dangerous, because Ansible use Jinja itself.

tomuben replied:
Fixed

tkilias commented on exasol_script_languages_developer_sandbox/templates/install_docker.ansible.yml:
```diff
@@ -0,0 +1,40 @@
+#- name: Install aptitude
```
Do we need this?

tomuben replied:
Removed

tkilias commented on exasol_script_languages_developer_sandbox/templates/install_docker.ansible.yml:
```diff
@@ -0,0 +1,40 @@
+#    state: latest
+#    update_cache: true
+
+- name: Install required system packages
```
Why do we have versions of this file?

tomuben replied:
Was a mistake. I removed it.

tkilias commented on test/ansible/slc_setup_test.yml:
```diff
@@ -0,0 +1,24 @@
+
+  tasks:
+
+    - name: add docker container to inventory
```
Do we need this , thought we wanted to remove it?

tomuben replied:
This is just a convenience to add the docker container to the repository. The docker container itself is started via the Docker SDK. 
I could manipulate the inventory by hand, but I found this method more safe and easy to maintain.

tomuben committed:
Fixes from review
1. Removed install_docker.ansible.yml
2. Removed aptitude code from install_docker.yml
3. Changed implementation in ansible_repository.py

tomuben committed:
fixed ansible_repository.py

tomuben committed:
Install APT dependencies

tomuben committed:
1. Changed apt dependency variable to a nicer format
2. Install github repo

tomuben committed:
1. Use ansible roles for different tasks
2. Implement setup of Jupyter lab
3. Added a integration test which validates execution of exaslct

tomuben committed:
fixed test_install_dependencies_jupyterlab

tomuben committed:
fixed test_install_dependencies_jupyterlab

tomuben committed:
Extended user guide

tomuben committed:
Use ansible variable order instead of custom default handling

tkilias commented on exasol_script_languages_developer_sandbox/lib/ansible/ansible_repository.py:
```diff
@@ -0,0 +1,72 @@
+
+class AnsibleResourceRepository(AnsibleRepository):
+    """
+    Represents a repository containing ansible files (roles, playbooks, taks, etc.).
```
```suggestion
    Represents a repository containing ansible files (roles, playbooks, tasks, etc.).
```

tkilias commented on exasol_script_languages_developer_sandbox/lib/run_install_dependencies.py:
```diff
@@ -0,0 +1,30 @@
+    are copied as flat copy to the dynamic working copy, too.
+    The playbook is indicated by variable ansible_run_context, which also might contain additional ansible variables.
+    """
+    # TODO: Awaiting release 5.0.0 of slc!!!
```
Create a ticket for the Todo

tomuben replied:
Done https://github.com/exasol/script-languages-developer-sandbox/issues/12

tkilias commented on exasol_script_languages_developer_sandbox/lib/run_setup_ec2_and_install_dependencies.py:
```diff
@@ -0,0 +1,59 @@
+    Runs setup of an EC2 instance and then installs all dependencies via Ansible.
+    Note that if an error occurs during the ansible installation, the EC2-machine is not removed immediately. This
+    gives you time to login into the machine and identify any setup issues.
+    You can stop the EC-2 machine by pressing Ctrl-C.
```
Nice, however this is not usable in automated usage

tomuben replied:
Yeah, the purpose of this is just debugging....

tkilias commented on exasol_script_languages_developer_sandbox/runtime/ansible/roles/docker/tasks/main.yml:
```diff
@@ -0,0 +1,32 @@
+    update_cache: true
+  become: "{{need_sudo}}"
+
+- name: adding docker users (for use without sudo)
```
```suggestion
- name: Adding docker users (for use without sudo)
```

tkilias commented on exasol_script_languages_developer_sandbox/runtime/ansible/roles/docker/tasks/main.yml:
```diff
@@ -0,0 +1,32 @@
+    name: "{{user_name}}"
+    append: yes
+    groups: docker
+  become: "{{need_sudo}}"
```
Is this because of docker. Maybe we should prepare the docker container with the same user and sudo settings as the AMI

tomuben replied:
Created an issue: #13

tkilias commented on exasol_script_languages_developer_sandbox/runtime/ansible/roles/jupyter/defaults/main.yml:
```diff
@@ -0,0 +1,12 @@
+jupyterlab:
+  ip: '*'
+  config: "{{user_home}}/.jupyter/jupyter_lab_config.py"
+  virtualenv: "{{user_home}}/jupytenv"
```
Strange env name

tomuben replied:
Fixed

tkilias commented on test/test_install_dependencies.py:
```diff
@@ -0,0 +1,78 @@
+        run_install_dependencies(AnsibleAccess(), host_infos=tuple(), ansible_run_context=ansible_run_context,
+                                 ansible_repositories=repos)
+        yield test_container, tmp_dir
+        test_container.exec_run(f"rm -rf {tmp_dir}/script-languages-release") ##script-languages-release is created as root
```
Comment is not completely clear

tomuben replied:
Added more details

tomuben committed:
1. Changed ansible variables for slc
2. Findings from review

tomuben committed:
Added some unit tests for the ansible lib

tkilias commented on test/test_ansible.py:
```diff
@@ -1,17 +1,132 @@
+        p = pathlib.Path(work_dir) / "slc_setup.yml"
+        assert p.exists()
+        p = pathlib.Path(work_dir) / "roles" / "script_languages" / "tasks" / "main.yml"
+        p.exists()
```
Missing assert

tomuben replied:
fixed

tomuben committed:
fixed test_ansible.py
